### PR TITLE
Replace adding '-isystem' with '-isystem-after'

### PR DESCRIPTION
**Pull request by irishrover**
This fixes issues when isystem already presents in command line,
for example Chromium.

### CODE REVIEW

**Comment by irishrover** [comment]
Typical command line for Chromium (Linux/Mac) contains something like

`-std=c++14 -stdlib=libc++ -fno-exceptions -fno-rtti -nostdinc++ 
-isystem../../buildtools/third_party/libc++/trunk/include 
-isystem../../buildtools/third_party/libc++abi/trunk/include 
`

So ``-isystem`` is heavily used there.

**Comment by irishrover** [comment]
Gentlt ping :-)

File: analyzer/codechecker_analyzer/buildlog/log_parser.py [hunk 1]
```diff
@@ -190,6 +190,7 @@
     '-[DIUF]',
     '-idirafter',
     '-isystem',
+    '-isystem-after',
```
[the remark this answers is not in the document]
Comment (irishrover) [reply analyzer/codechecker_analyzer/buildlog/log_parser.py]: Thanks, fixed.

**Comment by irishrover** [comment]
> In which case is it useful not to use the include paths of the original build environment primarily?

It's what I get when `-isystem` are used instead of `-isystem-after`:

`$> CodeChecker analyze --analyzers=clangsa ./compile_commands.json -j$(nproc) --verbose debug -o TMP_REPORTS/`:
```
clang --analyze -Qunused-arguments --analyzer-no-default-checks -Xclang -analyzer-opt-analyze-headers ...
 -x c++ --target=x86_64-unknown-linux-gnu -std=gnu++14 
-DUSE_UDEV -DUSE_AURA=1 -DUSE_GLIB=1 -DUSE_NSS_CERTS=1 -DUSE_X11=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE 
-D_LARGEFILE64_SOURCE -D_GNU_SOURCE -DCR_CLANG_REVISION="371202-8455294f-2" -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -DCOMPONENT_BUILD -D_LIBCPP_ABI_UNSTABLE -D_LIBCPP_ABI_VERSION=Cr -D_LIBCPP_ENABLE_NODISCARD -DCR_LIBCXX_REVISION=361348 
-DCR_SYSROOT_HASH=e7c53f04bd88d29d075bfd1f62b073aeb69cbe09 -D_DEBUG -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DGL_GLEXT_PROTOTYPES -DUSE_GLX -DUSE_EGL 
-DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_2_32 
-DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_26 -DWEBP_EXTERN=extern -DGTEST_API_= 
-DGTEST_HAS_POSIX_RE=0 -DGTEST_LANG_CXX11=1 -DGTEST_HAS_TR1_TUPLE=0 -DBORINGSSL_SHARED_LIBRARY -DSK_GL -DSK_HAS_PNG_LIBRARY -DSK_HAS_WEBP_LIBRARY -DSK_USER_CONFIG_HEADER="../../skia/config/SkUserConfig.h" -DSK_HAS_JPEG_LIBRARY -DSK_VULKAN_HEADER="../../skia/config/SkVulkanConfig.h" -DSK_VULKAN=1 -DSKIA_DLL -DSKCMS_API=__attribute__((visibility("default"))) -DSK_SUPPORT_GPU=1 -DSK_GPU_WORKAROUNDS_HEADER="gpu/config/gpu_driver_bug_workaround_autogen.h" -DVK_NO_PROTOTYPES -DU_USING_ICU_NAMESPACE=0 -DU_ENABLE_DYLOAD=0 -DUSE_CHROMIUM_ICU=1 -DICU_UTIL_DATA_IMPL=ICU_UTIL_DATA_FILE 
-DUCHAR_TYPE=uint16_t -DUNIT_TEST -I/home/zzz/Documents/chromium/src 
-I/home/zzz/Documents/chromium/src/out/Debug/gen 
-I/home/zzz/Documents/chromium/src/third_party/khronos 
-I/home/zzz/Documents/chromium/src/gpu 
-I/home/zzz/Documents/chromium/src/third_party/libwebp/src -I/home/zzz/Documents/chromium/src/third_party/googletest/custom -I/home/zzz/Documents/chromium/src/third_party/googletest/src/googletest/include -I/home/zzz/Documents/chromium/src/third_party/boringssl/src/include -I/home/zzz/Documents/chromium/src/third_party/mesa_headers -I/home/zzz/Documents/chromium/src/third_party/skia -I/home/zzz/Documents/chromium/src/third_party/vulkan/include -I/home/zzz/Documents/chromium/src/third_party/skia/third_party/vulkanmemoryallocator -I/home/zzz/Documents/chromium/src/third_party/vulkan/include -I/home/zzz/Documents/chromium/src/third_party/icu/source/common -I/home/zzz/Documents/chromium/src/third_party/icu/source/i18n -I/home/zzz/Documents/chromium/src/third_party/googletest/custom -I/home/zzz/Documents/chromium/src/third_party/googletest/src/googlemock/include -fno-strict-aliasing -fstack-protector -funwind-tables -fPIC -fcolor-diagnostics -fmerge-all-constants -fcrash-diagnostics-dir=../../tools/clang/crashreports -fcomplete-member-pointers -m64 -march=x86-64 -Wno-builtin-macro-redefined -D__DATE__= -D__TIME__= -D__TIMESTAMP__= -Wno-missing-field-initializers -Wno-unused-parameter -Wno-c++11-narrowing -Wno-unneeded-internal-declaration -Wno-undefined-var-template -Wno-ignored-pragma-optimize -Wno-implicit-int-float-conversion -Wno-xor-used-as-pow -Wno-c99-designator -Wno-reorder-init-list -Wno-final-dtor-non-final-class -fno-omit-frame-pointer -fvisibility=hidden -isystem../../build/linux/debian_sid_amd64-sysroot/usr/include/glib-2.0 
-isystem../../build/linux/debian_sid_amd64-sysroot/usr/lib/x86_64-linux-gnu/glib-2.0/include -Wno-inconsistent-missing-override -Wno-undefined-bool-conversion -Wno-tautological-undefined-compare -std=c++14 -fno-exceptions -fno-rtti -nostdinc++ 
-isystem../../buildtools/third_party/libc++/trunk/include 
-isystem../../buildtools/third_party/libc++abi/trunk/include 
--sysroot=../../build/linux/debian_sid_amd64-sysroot -fvisibility-inlines-hidden 
-isystem ../../build/linux/debian_sid_amd64-sysroot/usr/include/c++/6 
-isystem ../../build/linux/debian_sid_amd64-sysroot/usr/include/x86_64-linux-gnu/c++/6 
-isystem ../../build/linux/debian_sid_amd64-sysroot/usr/include/c++/6/backward 
-isystem /home/zzz/Documents/chromium/src/third_party/llvm-build/Release+Asserts/lib/clang/10.0.0/include 
-isystem ../../build/linux/debian_sid_amd64-sysroot/usr/include/x86_64-linux-gnu 
-isystem ../../build/linux/debian_sid_amd64-sysroot/usr/include /home/zzz/Documents/chromium/src/ui/gl/gl_mock.cc

In file included from /home/zzz/Documents/chromium/src/ui/gl/gl_mock.cc:5:
In file included from /home/zzz/Documents/chromium/src/ui/gl/gl_mock.h:12:
In file included from /home/zzz/Documents/chromium/src/testing/gmock/include/gmock/gmock.h:10:
In file included from /home/zzz/Documents/chromium/src/third_party/googletest/src/googlemock/include/gmock/gmock.h:59:
In file included from /home/zzz/Documents/chromium/src/third_party/googletest/src/googlemock/include/gmock/gmock-actions.h:44:
In file included from ../../buildtools/third_party/libc++/trunk/include/algorithm:643:
In file included from ../../buildtools/third_party/libc++/trunk/include/memory:653:
In file included from ../../buildtools/third_party/libc++/trunk/include/typeinfo:60:
In file included from ../../buildtools/third_party/libc++/trunk/include/exception:81:
In file included from ../../buildtools/third_party/libc++/trunk/include/cstdlib:85:
In file included from ../../buildtools/third_party/libc++/trunk/include/stdlib.h:97:
../../build/linux/debian_sid_amd64-sysroot/usr/include/c++/6/stdlib.h:38:12: error: no member named 'abort' in namespace 'std'
using std::abort;
      ~~~~~^
../../build/linux/debian_sid_amd64-sysroot/usr/include/c++/6/stdlib.h:39:12: error: no member named 'atexit' in namespace 'std'
using std::atexit;
      ~~~~~^
../../build/linux/debian_sid_amd64-sysroot/usr/include/c++/6/stdlib.h:40:12: error: no member named 'exit' in namespace 'std'
using std::exit;
      ~~~~~^
In file included from /home/zzz/Documents/chromium/src/ui/gl/gl_mock.cc:5:
In file included from /home/zzz/Documents/chromium/src/ui/gl/gl_mock.h:12:
In file included from /home/zzz/Documents/chromium/src/testing/gmock/include/gmock/gmock.h:10:
In file included from /home/zzz/Documents/chromium/src/third_party/googletest/src/googlemock/include/gmock/gmock.h:59:
In file included from /home/zzz/Documents/chromium/src/third_party/googletest/src/googlemock/include/gmock/gmock-actions.h:44:
In file included from ../../buildtools/third_party/libc++/trunk/include/algorithm:643:
In file included from ../../buildtools/third_party/libc++/trunk/include/memory:653:
In file included from ../../buildtools/third_party/libc++/trunk/include/typeinfo:60:
In file included from ../../buildtools/third_party/libc++/trunk/include/exception:81:
In file included from ../../buildtools/third_party/libc++/trunk/include/cstdlib:85:
In file included from ../../buildtools/third_party/libc++/trunk/include/stdlib.h:100:
In file included from ../../buildtools/third_party/libc++/trunk/include/math.h:303:
In file included from ../../build/linux/debian_sid_amd64-sysroot/usr/include/c++/6/math.h:36:
../../buildtools/third_party/libc++/trunk/include/cmath:313:9: error: no member named 'signbit' in the global namespace
using ::signbit;
      ~~^
../../buildtools/third_party/libc++/trunk/include/cmath:314:9: error: no member named 'fpclassify' in the global namespace
using ::fpclassify;
      ~~^
../../buildtools/third_party/libc++/trunk/include/cmath:315:9: error: no member named 'isfinite' in the global namespace
using ::isfinite;
      ~~^
../../buildtools/third_party/libc++/trunk/include/cmath:316:9: error: no member named 'isinf' in the global namespace
using ::isinf;
      ~~^
../../buildtools/third_party/libc++/trunk/include/cmath:317:9: error: no member named 'isnan' in the global namespace
using ::isnan;
      ~~^
../../buildtools/third_party/libc++/trunk/include/cmath:318:9: error: no member named 'isnormal' in the global namespace
using ::isnormal;
      ~~^
../../buildtools/third_party/libc++/trunk/include/cmath:319:9: error: no member named 'isgreater' in the global namespace
using ::isgreater;
      ~~^
../../buildtools/third_party/libc++/trunk/include/cmath:320:9: error: no member named 'isgreaterequal' in the global namespace
using ::isgreaterequal;
      ~~^
../../buildtools/third_party/libc++/trunk/include/cmath:321:9: error: no member named 'isless' in the global namespace
using ::isless;
      ~~^
../../buildtools/third_party/libc++/trunk/include/cmath:322:9: error: no member named 'islessequal' in the global namespace
using ::islessequal;
      ~~^
../../buildtools/third_party/libc++/trunk/include/cmath:323:9: error: no member named 'islessgreater' in the global namespace
using ::islessgreater;
      ~~^
../../buildtools/third_party/libc++/trunk/include/cmath:324:9: error: no member named 'isunordered' in the global namespace
using ::isunordered;
      ~~^
../../buildtools/third_party/libc++/trunk/include/cmath:325:9: error: no member named 'isunordered' in the global namespace
using ::isunordered;
      ~~^
../../buildtools/third_party/libc++/trunk/include/cmath:327:9: error: no member named 'float_t' in the global namespace
using ::float_t;
      ~~^
../../buildtools/third_party/libc++/trunk/include/cmath:328:9: error: no member named 'double_t' in the global namespace
using ::double_t;
      ~~^
../../buildtools/third_party/libc++/trunk/include/cmath:331:9: error: no member named 'abs' in the global namespace
using ::abs;
      ~~^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.

```

**Comment by irishrover** [comment]
Hi, @bruntib, do you have any other questions?

**Comment by bruntib** [comment]
Hi,

Sorry for being this slow. I'm trying to create a minimal example to reproduce the issue. I really would like to understand what exactly are the effects of this patch, because we wouldn't like to fail the analysis of our internal users. I can't see how the order of include paths can result a missing symbol during compilation. Maybe some `#include_next` issue somewhere?
By the way has this error always occurred in your analysis? If it went wrong at a certain point, then which commit caused the error? Me too I'm trying to find the answers on these questions as soon as possible.

**Comment by bruntib** [comment]
Hi,

This really seems a good fix. The error came in in this commit: https://github.com/Ericsson/codechecker/commit/ada3fde7c2d92f32369f482af1d5e033bedde6a9.

This was about not to exclude `include-fixed` directories from among implicit include paths. However, because of this the error should occur only if the build system of your analyzed project uses GCC. We'll run some more tests and merge the pull request soon.

Thanks

**Comment by bruntib** [comment]
Please, cherry-pick this commit to branch `release-v6.10.1`.
Thank you!

**Comment by bruntib** [comment]
Hi,

There still is a bug with this solution on some projects. Right now it seems to be a good solution if we use `-isystem-after` only for `include-fixed` directories (see https://github.com/Ericsson/codechecker/commit/ada3fde7c2d92f32369f482af1d5e033bedde6a9).

**Comment by irishrover** [comment]
Hi, bruntib!

Thanks for your assistance. 

> Please, cherry-pick this commit to branch release-v6.10.1.

Seems it has been done already?

> This really seems a good fix. The error came in in this commit: ada3fde.
> By the way has this error always occurred in your analysis? If it went wrong at a certain point, then which commit caused the error? 

No, in fact I 've been facing with this all the times, at least for the last 1.5 year when trying to analyze Chromium sources. And this patch is the last change that prevents Chromium from being analyzed by the CodeChecker master branch out-of-the-box.

> Maybe some #include_next issue somewhere?

Yes, I think so, but I still don't know how reproduce in a small sample. `#include_next` can appear as Chromium uses both system and custom headers, like:

-isystem../../build/linux/debian_sid_amd64-sysroot/usr/lib/x86_64-linux-gnu/glib-2.0/include 
-isystem../../buildtools/third_party/libc++/trunk/include 
-isystem../../buildtools/third_party/libc++abi/trunk/include 

I can share an Ubuntu VMWare image with the CodeChecker and Chromium rolled. So you can play with it.  Or I can provide you an instruction how to build Chromium manually on your own system (note, it takes few hours to download, setup and build).

**Comment by bruntib** [comment]
Hi,

No, it has not been cherry-picked to the other branch, but we discussed that in case of external contributors we can do it ourselves.

However, unfortunately this is not the final solution yet, since with this change our projects don't compile :) So we'll need some deeper understanding on the issue.

**Comment by irishrover** [comment]
Can I help you with that?
Do you need my VMWare image with Chromium + CodeChecker?

**Comment by bruntib** [comment]
Yes, it would be nice. Thank you!

**Comment by bruntib** [comment]
And could you also send us an exact example on a build command from your compilation database JSON and the corresponding clang analyzer invocation from the debug logs? (Actually you have sent the latter one earlier)

**Comment by bruntib** [comment]
Hi,

We have one more silver bullet for the issue :) Could you try this patch on your project, please? https://github.com/Ericsson/codechecker/pull/2344

This reverts the `-isystem-after` change (because it resulted much more compilation errors at us) and takes -nostdinc into account when determining implicit include paths.

Thank you!

**Comment by irishrover** [comment]
> Hi,
> 
> We have one more silver bullet for the issue :) Could you try this patch on your project, please? #2344
> 
> This reverts the `-isystem-after` change (because it resulted much more compilation errors at us) and takes -nostdinc into account when determining implicit include paths.
> 
> Thank you!

Oh, nice, thanks! 
I'll do it in few hours.

**Comment by irishrover** [comment]
> Hi,
> 
> We have one more silver bullet for the issue :) Could you try this patch on your project, please? #2344
> 
> This reverts the `-isystem-after` change (because it resulted much more compilation errors at us) and takes -nostdinc into account when determining implicit include paths.
> 
> Thank you!

Cool, it helps! Chromium is analyzed without any issues with this change!

**Comment by bruntib** [comment]
No, It is not necessary, thank you. The remainder has to be fixed for our analyzed projects which is by the way reproducible in some open source codes too.

Thanks again for your help.

**Pull request merged**